recipes-kernel/linux/files/patches-6.1/0081-iio-light-Add-support-for-everlight-pmd16d17-sensor.patch

```diff
@@ -0,0 +1,408 @@
++#define PM16D17_PS_OFFSET_CANCEL_L	0x14
++#define PM16D17_PS_OFFSET_CANCEL_H	0x15
++#define PM16D17_DEV_ID			0x18
++
++#define DEVICE_ID			0x11
++
++#define ENABLE_PS_FUNCTION		BIT(3)
++#define PS_GAIN_MASK			GENMASK(7, 6)
++#define PS_ITIME_MASK			GENMASK(5, 3)
++#define PS_WTIME_MASK			GENMASK(2, 0)
++#define OFFSET_CANCEL_ENABLE		BIT(7)
++#define PS_OFFSET_CANCEL_LSB_MASK	GENMASK(7, 0)
++#define PS_OFFSET_CANCEL_MSB_MASK	GENMASK(15, 8)
++
++enum {
++	PITIME_0_POINT_4_MS = (0 << 3),
++	PITIME_0_POINT_8_MS = (1 << 3),
++	PITIME_1_POINT_6_MS = (2 << 3),
++	PITIME_3_POINT_2_MS = (3 << 3),
++	PITIME_6_POINT_3_MS = (4 << 3),
++	PITIME_12_POINT_6_MS = (5 << 3),
++	PITIME_25_POINT_2_MS = (6 << 3),
++};
++
++enum {
++	PWTIME_12_POINT_5_MS = 0,
++	PWTIME_25_MS,
++	PWTIME_50_MS,
++	PWTIME_100_MS,
++	PWTIME_200_MS,
++	PWTIME_400_MS,
++	PWTIME_800_MS,
++	PWTIME_1600_MS,
++};
++
++struct pm16d17_data {
++	struct i2c_client *client;
++	struct regmap *regmap;
++	struct mutex lock;
++};
++
++static const struct regmap_config pm16d17_regmap_config = {
++	.name = PM16D17_REGMAP_NAME,
++	.reg_bits = 8,
++	.val_bits = 8,
++	.cache_type = REGCACHE_NONE,
++};
++
++static const struct iio_chan_spec pm16d17_channels[] = {
++	{
++		.type = IIO_LIGHT,
++		.indexed = 1,
++		.channel = 0,
++		.scan_index = 0,
++		.info_mask_separate = BIT(IIO_CHAN_INFO_RAW),
++	}
++};
++
++static inline int pm16d17_write_reg(struct pm16d17_data *data,
++				    unsigned int reg,
++				    unsigned int value)
++{
++	return regmap_write(data->regmap, reg, value);
++}
++
++static inline unsigned int pm16d17_read_reg(struct pm16d17_data *data,
++					    unsigned int reg,
++					    unsigned int *reg_val)
++{
++	return regmap_read(data->regmap, reg, reg_val);
++}
++
++static int pm16d17_read_raw(struct iio_dev *indio_dev,
++			    struct iio_chan_spec const *chan,
++			    int *val, int *val2, long mask)
++{
++	struct pm16d17_data *data = iio_priv(indio_dev);
++	unsigned int ps_data_l;
++	unsigned int ps_data_h;
++	uint16_t ps_data;
++	int ret = -EINVAL;
++
++	mutex_lock(&data->lock);
++
++	switch (mask) {
++	case IIO_CHAN_INFO_RAW:
++		switch (chan->type) {
++		case IIO_LIGHT:
++			ret = pm16d17_read_reg(data, PM16D17_PS_DATA_L, &ps_data_l);
++			if (ret < 0)
++				goto fail;
++
++			ret = pm16d17_read_reg(data, PM16D17_PS_DATA_H, &ps_data_h);
++			if (ret < 0)
++				goto fail;
++
++			ps_data = (ps_data_h << 8) | ps_data_l;
++
++			dev_dbg(&data->client->dev, "PS data: %x\n", ps_data);
++
++			*val = ps_data;
++			ret = IIO_VAL_INT;
++			break;
++		default:
++			break;
++		}
++	default:
++		break;
++	}
++fail:
++	mutex_unlock(&data->lock);
++
++	return ret;
++}
++
++static int pm16d17_write_raw(struct iio_dev *indio_dev,
++			     struct iio_chan_spec const *chan,
++			     int val, int val2, long mask)
++{
++	return 0;
++}
++
++
++static int pm16d17_write_raw_get_fmt(struct iio_dev *indio_dev,
++				     struct iio_chan_spec const *chan,
++				     long mask)
++{
++	return 0;
```

Comment:
Is that correct? I'm not finding other IIO drivers doing that. I rather think you are supposed to return some `IIO_VAL_*` constant.

Reply:
Fixed in #509, the write interface is removed, since we don't need it at this moment. Also checked some similar drivers, write interface is not a must provided one.